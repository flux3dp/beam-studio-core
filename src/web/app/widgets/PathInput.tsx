import * as React from 'react';
import classNames from 'classnames';
import electronDialogs from 'app/actions/electron-dialogs';
import fs from 'fs';
import { IFileFilter } from 'interfaces/IElectron';

const { useEffect, useRef, useState } = React;

export enum InputType {
  FILE = 0,
  FOLDER = 1,
  BOTH = 2,
}

const propertiesMap = {
  0: ['openFile'],
  1: ['openDirectory', 'createDirectory', 'promptToCreate'],
  2: ['openFile', 'openDirectory', 'createDirectory', 'promptToCreate'],
};

interface Props {
  buttonTitle: string,
  className: string
  defaultValue: string,
  forceValidValue: boolean,
  type: InputType,
  getValue: (val: string, isValid: boolean) => void,
  // eslint-disable-next-line react/require-default-props
  onBlur?: () => void,
  // eslint-disable-next-line react/require-default-props
  filters?: IFileFilter[],
}

const PathInput = ({
  buttonTitle,
  className,
  defaultValue,
  getValue,
  forceValidValue = true,
  onBlur = () => { },
  type = InputType.FILE,
  filters = [],
}: Props) => {
  const [displayValue, setDisplayValue] = useState(defaultValue);
  const [savedValue, setSavedValue] = useState(defaultValue);
  const inputEl = useRef(null);

  useEffect(() => {
    setDisplayValue(defaultValue);
    setSavedValue(defaultValue);
  }, [defaultValue]);

  const validateValue = (val: string) => {
    if (fs.existsSync(val)) {
      if (type === InputType.BOTH) return true;
      const stat = fs.lstatSync(val);
      return (type === InputType.FILE && stat.isFile())
        || (type === InputType.FOLDER && stat.isDirectory());
    }
    return false;
  };

  const updateValue = () => {
    const isValid = validateValue(displayValue);
    if (!forceValidValue || isValid) {
      if (displayValue !== savedValue) {
        setSavedValue(displayValue);
        if (getValue) {
          getValue(displayValue, isValid);
        }
      }
    } else {
      setDisplayValue(savedValue);
    }
  };

  const handleBlur = () => {
    updateValue();
    if (onBlur) {
      onBlur();
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const target = e.target as HTMLInputElement;
    setDisplayValue(target.value);
  };

  const handleKeyUp = (e: React.KeyboardEvent) => {
    // backspace somehow does not trigger onchange event
    const target = e.target as HTMLInputElement;
    if (target.value !== displayValue) {
      setDisplayValue(target.value);
    }
  };

  const setValueFromDialog = async () => {
    const properties = propertiesMap[type];
    const option = {
      properties,
      filters,
      defaultPath: savedValue,
    };
    const { filePaths, canceled } = await electronDialogs.showOpenDialog(option);
    if (!canceled) {
      const isValid = validateValue(filePaths[0]);
      if (!forceValidValue || isValid) {
        setSavedValue(filePaths[0]);
        setDisplayValue(filePaths[0]);
        if (getValue) {
          getValue(filePaths[0], isValid);
        }
      }
    }
  };

  const openDialogButton = (
    <div className="dialog-btn" title={buttonTitle} onClick={setValueFromDialog}>
      <img src="img/right-panel/icon-import.svg" />
    </div>
  );

  return (
    <div className={classNames('path-input', className)}>
      <input
        type="text"
        value={displayValue}
        onBlur={handleBlur}
        onChange={handleChange}
        onKeyUp={handleKeyUp}
        ref={inputEl}
      />
      {openDialogButton}
    </div>
  );
};

export default PathInput;
