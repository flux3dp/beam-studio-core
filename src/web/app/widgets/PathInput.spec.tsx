/* eslint-disable import/first */
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const mockShowOpenDialog = jest.fn();
jest.mock('implementations/dialog', () => ({
  showOpenDialog: mockShowOpenDialog,
}));

const mockExists = jest.fn();
const mockIsFile = jest.fn();
const mockIsDirectory = jest.fn();
jest.mock('implementations/fileSystem', () => ({
  exists: mockExists,
  isFile: mockIsFile,
  isDirectory: mockIsDirectory,
}));

import PathInput from './PathInput';

describe('test PathInput', () => {
  test('should render correctly', async () => {
    const mockGetValue = jest.fn();
    const wrapper = shallow(<PathInput
      buttonTitle="Choose Folder"
      className="with-error"
      defaultValue="defaultFolder"
      forceValidValue={false}
      getValue={mockGetValue}
      type={1}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    await mockShowOpenDialog.mockResolvedValue({
      filePaths: [],
      canceled: true,
    });
    wrapper.find('div.dialog-btn').simulate('click');
    await (() => expect(mockShowOpenDialog).toHaveBeenCalledTimes(1));
    await (() => expect(mockShowOpenDialog).toHaveBeenNthCalledWith(1, {
      properties: ['openDirectory', 'createDirectory', 'promptToCreate'],
      defaultPath: 'defaultFolder',
    }));
    await (() => expect(mockGetValue).not.toHaveBeenCalled());
    await expect(toJson(wrapper)).toMatchSnapshot();

    await mockShowOpenDialog.mockResolvedValue({
      filePaths: ['myDocuments'],
      canceled: false,
    });
    mockExists.mockReturnValue(false);
    wrapper.find('div.dialog-btn').simulate('click');
    await (() => expect(mockShowOpenDialog).toHaveBeenCalledTimes(2));
    await (() => expect(mockShowOpenDialog).toHaveBeenNthCalledWith(2, {
      properties: ['openDirectory', 'createDirectory', 'promptToCreate'],
      defaultPath: 'defaultFolder',
    }));
    await (() => expect(mockExists).toHaveBeenCalledTimes(1));
    await (() => expect(mockExists).toHaveBeenNthCalledWith(1, 'myDocuments'));
    await (() => expect(mockGetValue).toHaveBeenCalledTimes(1));
    await (() => expect(mockGetValue).toHaveBeenNthCalledWith(1, 'myDocuments', false));
    await expect(toJson(wrapper)).toMatchSnapshot();
  });
});
