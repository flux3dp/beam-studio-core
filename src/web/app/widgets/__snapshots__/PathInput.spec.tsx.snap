// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PathInput should render correctly 1`] = `
<div
  className="path-input with-error"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="defaultFolder"
  />
  <div
    className="dialog-btn"
    onClick={[Function]}
    title="Choose Folder"
  >
    <img
      src="img/right-panel/icon-import.svg"
    />
  </div>
</div>
`;

exports[`test PathInput should render correctly 2`] = `
<div
  className="path-input with-error"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="defaultFolder"
  />
  <div
    className="dialog-btn"
    onClick={[Function]}
    title="Choose Folder"
  >
    <img
      src="img/right-panel/icon-import.svg"
    />
  </div>
</div>
`;

exports[`test PathInput should render correctly 3`] = `
<div
  className="path-input with-error"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="myDocuments"
  />
  <div
    className="dialog-btn"
    onClick={[Function]}
    title="Choose Folder"
  >
    <img
      src="img/right-panel/icon-import.svg"
    />
  </div>
</div>
`;
